Add ListFiles function

SmilingNavern opened the pull request.

SmilingNavern commented:
I am not sure about this approach. Glob() and ReadDir() behaves differently in terms of output. So i provide relative path to files according to args.

SmilingNavern commented on sources.go:
```diff
@@ -17,6 +19,33 @@ func File(name string) *Pipe {
+func ListFiles(path string) *Pipe {
+	p := NewPipe()
+
+	if strings.Contains(path, "*") {
```
I am really not familiar with windows programming but i am going to look into this. Maybe i would need some help)

bitfield replied:
I think PR #5 handles this correctly, if you want to compare notes with that one.

SmilingNavern commented:
I am going to fix this one then.

SmilingNavern commented:
@bitfield i've updated my PR, thanks for such long delay. I've used ContainsAny and Readme parts from #5 and updated them a little. @abtin Check it please as well.

bitfield commented:
Great! Could you check that you've addressed all the code review comments so far?

SmilingNavern commented:
@bitfield i think i have addressed them all except one about Echo(), because i am not sure if this necessary and i am not sure about quality of Readme section, because i am bad in english and readme as well

SmilingNavern commented:
Actually i thought that ListFiles maybe not a great name. It's more like ListDir(). But i didnt change that because i am not sure. The logic of code is more similiar to something that readdir() do.

bitfield commented:
Yes, wherever you're creating a pipe from a string, `Echo()` is the thing to use. That means if we want to change the implementation from a `strings.Reader` to something else later, we can.

Don't worry about the Readme text, I'll copyedit that stuff and tweak the wording.

I'm not crazy about the ListFiles() name either, but I can't think of anything better at the moment. `Ls()` is too Unix-specific. `Files()` could be confused with what `Concat()` does—read the contents of the files, not just get the names.

This is where an example program really helps; it shows you how the name of the method will read when combined with others. Let's see if we can come up with an example program which gets a list of files and processes the names in some way.

SmilingNavern commented:
Okay, i am not sure about name really. Let's stick to `ListFiles()` for now.

I will rewrite to Echo() and think about example.

SmilingNavern commented:
I've fixed PR for using Echo() to create Pipe.

SmilingNavern commented:
I've added example. But i think we need more function for files to provide more good examples. Maybe something like `RemoveFile()` or `Gzip()`/`Tar()`

SmilingNavern commented:
Right now it's not really convenient to do something with this list of path except concating it to match something. I am not sure that `EachLine()` is convenient function for this type of task.

abtin commented:
Actually I think an iterator function (e.g. Iter, ForEach) can be very
useful.

On Tue, Jul 9, 2019, 7:11 PM Alexander Danilov <notifications@github.com>
wrote:

> Right now it's not really convenient to do something with this list of
> path except concating it to match something. I am not sure that EachLine()
> is convenient function for this type of task.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitfield/script/pull/2?email_source=notifications&email_token=AADLYHFY5ANJRCTPM6CQBM3P6ULLPA5CNFSM4HWJPYNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZRZQDY#issuecomment-509843471>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADLYHEH6VNIGMQ5TKTKQTTP6ULLPANCNFSM4HWJPYNA>
> .
>

SmilingNavern commented:
Yeap, i agree. But i am not sure how to implement it with Pipe concept.

SmilingNavern commented:
@bitfield hello, are there anything you want me to improve in this PR? I would be glad to hear any additional feedback.

bitfield commented:
I'll review this as soon as I get a chance! It's looking very good!

bitfield commented on README.md:
```diff
@@ -375,6 +376,20 @@ fmt.Println(output)
 
+## ListFiles
+
+`ListFiles()` Creates a pipe with list of files under the given path or glob.
```
How about something like this:

`ListFiles()` lists files, like Unix [`ls`](examples/ls/main.go). It creates a pipe containing all files and directories matching the supplied path specification, one per line. This can be the name of a directory (`/path/to/dir`), the name of a file (`/path/to/file`), or a _glob_ (wildcard expression) conforming to the syntax accepted by [filepath.Match](https://golang.org/pkg/path/filepath/#Match) (`/path/to/*`).

```go
p := script.ListFiles("/tmp/*.php")
files, err := p.String()
if err != nil {
	log.Fatal(err)
}
fmt.Println("found suspicious PHP files in /tmp:")
fmt.Println(files)
```

bitfield replied:
Don't forget to add `ls` to the list of examples at the end of the README.

SmilingNavern replied:
I've copied your exact description to readme. I am really bad at this, so if there are any issues with readme/docs it would be great if you can fix them.

bitfield commented on examples/grep-multiple-files/data/1.log:
```diff
@@ -0,0 +1 @@
+hello
```
I think maybe let's drop this second example for now—it doesn't really tell us much more than we already know from the `ls` and `grep` examples.

SmilingNavern replied:
I've dropped it

bitfield commented on examples/ls/main.go:
```diff
@@ -0,0 +1,22 @@
+)
+
+func main() {
+	script.ListFiles(os.Args[1]).Stdout()
```
We need to handle the case where there are no command-line arguments.

bitfield replied:
Also, I think we need to handle the special cases of `.` and `..`, which Unix `ls` does.

SmilingNavern replied:
i've fixed ls example to hide files starting with single dot. But i don't understand how do you want me to handle double dot? What is special with it?

bitfield commented on sources.go:
```diff
@@ -17,6 +19,33 @@ func File(name string) *Pipe {
 	return p.WithReader(f)
 }
 
+// ListFiles contains list of files in `path` as string per line
```
Don't forget we need a full doc comment here, explaining everything the method does.

bitfield commented on sources.go:
```diff
@@ -15,6 +17,27 @@ func Args() *Pipe {
+	if err != nil {
+		return NewPipe().WithError(err)
+	}
+	var fileNames []string
```
Let's use `make` to set the required capacity of this slice when we create it. That's more efficient than creating it with the default capacity and allowing `append` to repeatedly resize it.

SmilingNavern replied:
Fixed this as well

bitfield commented on sources_test.go:
```diff
@@ -26,6 +27,42 @@ func TestFile(t *testing.T) {
+func TestListFilesMultipleFiles(t *testing.T) {
+	t.Parallel()
+	dir := "testdata/multiple_files"
+	want := fmt.Sprintf("%s/1.txt\n%s/2.txt\n%s/3.tar.zip", dir, dir, dir)
```
This is a little confusing; let's write it out in full.

SmilingNavern replied:
Could you please explain what do you mean by full?

bitfield commented on sources_test.go:
```diff
@@ -26,6 +27,42 @@ func TestFile(t *testing.T) {
+	t.Parallel()
+	p := ListFiles("nonexistentpath")
+	if p.Error() == nil {
+		t.Errorf("want error status on listing non-existent path, but got nil")
```
This should be `t.Error`.

SmilingNavern replied:
Fixed this one

bitfield commented on sources_test.go:
```diff
@@ -26,6 +27,42 @@ func TestFile(t *testing.T) {
+func TestListFilesGlob(t *testing.T) {
+	t.Parallel()
+	dir := "testdata/multiple_files"
+	want := fmt.Sprintf("%s/1.txt\n%s/2.txt", dir, dir)
```
Same again, let's write out the expected string in full.

bitfield commented on sources_test.go:
```diff
@@ -26,6 +27,42 @@ func TestFile(t *testing.T) {
+		t.Fatal(err)
+	}
+	if want != got {
+		t.Errorf("In %q: want %q, got %q", dir, want, got)
```
Both `want` and `got` already contain the `dir` component, so we don't need to repeat it here.

SmilingNavern replied:
Fixed this one

bitfield commented:
@SmilingNavern how are you getting along? Sorry I seem to have given you a very long list of changes to make here. It might actually be more efficient for me to make them myself on your branch, then we can agree any final steps that need to happen before merging. What do you think?

If you'd prefer to own the whole thing from start to finish, that's fine too—just let me know.

SmilingNavern commented:
@bitfield i am going to finish this on weekend, so, please give me a time:) i was busy with work before.

SmilingNavern commented:
I've fixed most of issues with my code based on your feedback. If there are any more of them feel free to update current branch and pull request, so we won't delay this PR any longer:)

thanks for your good feedback and sorry if my english isn't great:)

bitfield commented:
Wonderful, thanks! I’ll review this as soon as I can.

bitfield commented:
I think we're about there! If you approve my final tweaks, I'll go ahead and merge this. Great job!

SmilingNavern commented:
Yeah, let's merge this

bitfield commented:
Thanks again for a great contribution! It's much appreciated.
